### ISSUE TITLE
Memory leak during training with standard DataLoader coupled with WebDataset dataloader

### ISSUE THREAD

**Issue by vishaal27** [webdataset/webdataset#357]
I am training CLIP models on a custom data mixture. This means I have two dataloaders: one is a standard dataloader with ImageFolder dataset objects and the other one is a webdataset loader as used in open_clip.
I am running into an issue where I notice that my CPU memory usage increases but is nothing drastic (way below my RAM limits). However, I noticed that the swap space usage increases exorbitantly leading to my jobs getting killed. Here are some plots showing the problem.

The green lines below showcase a working training run where everything is smooth (this is not with a webdataset loader but standard dataloader with ImageFolder datasets). However, in the red lines, I have both the image data-loader and a webdataset loader loading data together, and then I cycle through them repeatedly. At the point where the red lines stop, my training job gets killed. From the CPU RAM plots, nothing untoward seems to be happening, there is an initial spike but then things stabilise, similar to the green line. However, when I track the swap space (not shown in the plots), it seems to be rapidly going down, until it becomes 0, and then a few seconds later the job crashes. 

<img width="1543" alt="Screenshot 2024-05-20 at 4 47 08 PM" src="https://github.com/webdataset/webdataset/assets/21314772/f8c38910-802a-42b3-a151-1a207f6f625b">
<img width="1530" alt="Screenshot 2024-05-20 at 4 47 15 PM" src="https://github.com/webdataset/webdataset/assets/21314772/b1ce016c-e6ee-4a3f-bd6e-bfa1ce2b9951">

The data-loading code for the green line looks something like this:
```python
# Start Training.
for ind, data in enumerate(train_loader):
    batch_size = len(data["targets"])
```
and for the red line looks like:
```python
dl = get_wds_dataset(
    './data/{0000..0090.tar}',
    preprocess_img=transform,
    preprocess_label=lambda x: x,
    is_train=True,
    epoch=0,
    batch_size=512,
    num_samples=1_000_000,
    resampled=False,
    seed=0,
    workers=2,
)

# Start Training.
for ind, data in enumerate(train_loader):
    batch_size = len(data["targets"])
```
where `get_wds_dataset` is the function from here: https://github.com/mlfoundations/open_clip/blob/2e8de8312ea6185df7bc24a73b19a195a801a9fc/src/training/data.py#L328

I have tried running the script with `workers=0` and it still fails and gets killed after a while. Do you have any ideas on how to resolve this?

**Comment by vishaal27** [comment]
update: this was because of the `itertools.cycle` function leading to memory leaks (https://github.com/pytorch/pytorch/issues/23900). Without the `cycle` function everything works smoothly.

**Comment by volcverse** [comment]
Hi vishaal, I met the same problem of increasing memory as training goes on. And I identify the problem lies in the WebLoader. 

I tried a simple code snipet as shown below but found that the memory leakage still exists.
I wonder whether those functions that initialize tensors cause the leakage, like torch.Tensor, torch.zeros, torch.ones. Do you have any ideas? Any response will be greatly appreciated :)

```
def collate_fn(sample):
    test_zero_image = torch.zeros((3, 768, 768))
    return {
        'image': test_zero_image
    }


train_dataset = wds.WebDataset(['test.tar'], repeat=True, resampled=True, nodesplitter=wds.split_by_worker).decode('torch').map(collate_fn)

train_dataloader = wds.WebLoader(train_dataset, num_workers=4, batch_size=16, pin_memory=True, persistent_workers=True, prefetch_factor=4)

for i, b in enumerate(train_dataloader):
    print(f'test batch {i}')
```

**Comment by vishaal27** [comment]
Hi, for me the issue was with the `cycle` part of the loader, where this was creating copies of tensors unnecessarily and eating my ram. However, it seems that you are not using `cycle` right? However, I see that in your train_dataset signature, there is a `repeat=True`, perhaps this might be causing the issue? Could you check without using repeat.

**Comment by tmbdev** [comment]
Sorry to have been sitting on this for so long.

The `repeat=` argument doesn't use `itertools.cycle`, it simply sets `self.repetitions`; the main iterator is this:

```
    def iterator(self):
        """Create an iterator through the entire dataset, using the given number of repetitions."""
        for _ in range(self.repetitions):
            count = 0            
            for sample in self.iterator1():
                yield sample
                count += 1
            if count == 0:
                # if the dataset is empty, don't keep looping
                break
```

Generally speaking, WebDataset does not hold on to any data anywhere; it is just a sequences of iterators.

The only buffering or storage is explicitly inside the `shuffle` function.

**Comment by noamsgl** [comment]
Hi @tmbdev thanks for your contribution!

> The only buffering or storage is explicitly inside the shuffle function.

I'm wondering is it possible to shuffle the keys before loading the data into memory?
My pipeline is initialized like this:

```
train_loader = wds.DataPipeline(
        wds.SimpleShardList(train_paths),
        wds.shuffle(config.get_int("data.shard_shuffle_buffer_size")),
        wds.tarfile_to_samples(),
        wds.shuffle(config.get_int("data.sample_shuffle_buffer_size")),
        wds.decode("l"),
        wds.to_tuple("segment.npy label.npy __key__"),
        wds.map_tuple(torch.from_numpy, torch.from_numpy, int),
        wds.batched(config.training.train_batch_size),
    )
``` 

but this fills a buffer of in-memory undecoded byte numpy arrays.
Is it possible to shuffle the samples within each shard but without loading the data to a buffer?
